sgkit_vcf/vcf_reader.py

```diff
@@ -0,0 +1,286 @@
+                    f"Multiple input regions must be a sequence of sequence of strings: {regions}"
+                )
+            input_regions = regions
+
+    assert len(inputs) == len(input_regions)
+
+    tempdir = Path(tempfile.mkdtemp(prefix="vcf_to_zarr_"))
+
+    datasets = []
+    parts = []
+    for i, input in enumerate(inputs):
+        filename = Path(input).name
+        input_region_list = input_regions[i] or [None]  # type: ignore
+        for r, region in enumerate(input_region_list):
+            part = tempdir / filename / f"part-{r}.zarr"
+            parts.append(part)
+            ds = dask.delayed(vcf_to_dataset)(
+                input,
+                output=part,
+                region=region,
+                chunk_length=chunk_length,
+                chunk_width=chunk_width,
+            )
+            datasets.append(ds)
+    datasets = dask.compute(*datasets)
+
+    # Ensure Dask task graph is efficient, see https://github.com/dask/dask/issues/5105
+    dask.config.set({"optimization.fuse.ave-width": 50})
+
+    # Combine the datasets into one
+    ds = xr.concat(datasets, dim="variants", data_vars="minimal")  # type: ignore[no-untyped-call, no-redef]
+    ds: xr.Dataset = ds.chunk({"variants": chunk_length, "samples": chunk_width})  # type: ignore
+
+    # Set variable length strings to fixed length ones to avoid xarray/conventions.py:188 warning
+    # (Also avoids this issue: https://github.com/pydata/xarray/issues/3476)
+    max_variant_id_length = max(ds.attrs["max_variant_id_length"] for ds in datasets)
+    max_variant_allele_length = max(
+        ds.attrs["max_variant_allele_length"] for ds in datasets
+    )
+    ds["variant_id"] = ds["variant_id"].astype(f"S{max_variant_id_length}")
+    ds["variant_allele"] = ds["variant_allele"].astype(f"S{max_variant_allele_length}")
+    del ds.attrs["max_variant_id_length"]
+    del ds.attrs["max_variant_allele_length"]
+
+    delayed = ds.to_zarr(output, mode="w", compute=False)
+    # delayed.visualize()
+    delayed.compute()
+
+    # Delete intermediate files from temporary directory
+    shutil.rmtree(tempdir)
+
+
+def vcf_to_zarr(
+    input: Union[PathType, Sequence[PathType]],
+    output: PathType,
+    *,
+    regions: Union[None, Sequence[str], Sequence[Optional[Sequence[str]]]] = None,
+    chunk_length: int = 10_000,
+    chunk_width: int = 1_000,
+) -> None:
+    if (isinstance(input, str) or isinstance(input, Path)) and (
+        regions is None or isinstance(regions, str)
+    ):
+        vcf_to_zarr_sequential(
+            input,
+            output,
+            region=regions,
+            chunk_length=chunk_length,
+            chunk_width=chunk_width,
+        )
+    else:
+        vcf_to_zarr_parallel(
+            input,
+            output,
+            regions=regions,
+            chunk_length=chunk_length,
+            chunk_width=chunk_width,
+        )
+
+
+def count_variants(path: PathType, region: Optional[str] = None) -> int:
```

Comment:
Is there ever a way to short-circuit this with data in a tabix index?

I'm looking at https://www.biostars.org/p/166414/#166448:

> The htslib implementation of tabix should have this information in dummy bins, I think.

Assuming some htslib-based thing built the index and that regions are often None, perhaps that's a worthwhile optimization.

Reply:
This isn't a public function, it's there for testing the tabix index count is the same as counting directly. Perhaps this should live in the tests?

Reply:
Ah I see, nm then.